File: analysis/engine/roster.py
"""Team rosters, depth charts, per-play lineups (spec §14, §15).

Depth-chart *order* uses `overall` (allowed — it only ranks players into
roles, §3/§15). Play *outcomes* use the individual skill attributes via the
rating layer (engine/ratings.py), never `overall`.
"""

from __future__ import annotations

import functools
import json
from pathlib import Path

_POOL = Path(__file__).resolve().parents[2] / "data" / "players.local.json"

OFF_SLOTS = ("QB1", "RB1", "WR1", "WR2", "WR3", "TE1", "LT", "LG", "C", "RG", "RT")
DEF_SLOTS_BASE = ("EDGE1", "EDGE2", "DT1", "DT2", "ILB1", "ILB2", "CB1", "CB2", "S1", "S2")
DEF_SLOTS_NICKEL = ("EDGE1", "EDGE2", "DT1", "DT2", "ILB1", "CB1", "CB2", "CB3", "S1", "S2")


@functools.lru_cache(maxsize=1)
def load_pool() -> dict[str, list[dict]]:
    players = json.loads(_POOL.read_text())
    by_team: dict[str, list[dict]] = {}
    for p in players:
        t = p.get("nfl_team", "")
        if not t or t == "FA":
            continue
        by_team.setdefault(t, []).append(p)
    return by_team


class Roster:
    def __init__(self, team: str, players: list[dict]):
        self.team = team
        self.depth: dict[str, list[dict]] = {}
        for p in players:
            self.depth.setdefault(p["position"], []).append(p)
        for pos in self.depth:
            self.depth[pos].sort(key=lambda x: x.get("overall", 0), reverse=True)
        self._off_cache: dict | None = None
        self._def_cache: dict[bool, dict] = {}

    def _nth(self, pos: str, i: int, out: set | None = None) -> dict | None:
        d = self.depth.get(pos, [])
        if out:
            avail = [p for p in d if p["id"] not in out]
            if avail:
                d = avail
        return d[i] if i < len(d) else (d[-1] if d else None)

    def offense(self, out: set | None = None) -> dict[str, dict]:
        """`out` = player ids to treat as unavailable (in-game injuries)."""
        if out:
            return self._build_offense(out)
        if self._off_cache is None:
            self._off_cache = self._build_offense()
        return self._off_cache

    def defense(self, nickel: bool = False, out: set | None = None) -> dict[str, dict]:
        if out:
            return self._build_defense(nickel, out)
        if nickel not in self._def_cache:
            self._def_cache[nickel] = self._build_defense(nickel)
        return self._def_cache[nickel]

    def _build_offense(self, out: set | None = None) -> dict[str, dict]:
        ot = self.depth.get("OT", [])
        og = self.depth.get("OG", [])
        if out:
            ot = [p for p in ot if p["id"] not in out]
            og = [p for p in og if p["id"] not in out]
        return {
            "QB1": self._nth("QB", 0, out),
            "RB1": self._nth("RB", 0, out),
            "WR1": self._nth("WR", 0, out), "WR2": self._nth("WR", 1, out), "WR3": self._nth("WR", 2, out),
            "TE1": self._nth("TE", 0, out),
            "LT": ot[0] if ot else None, "RT": ot[1] if len(ot) > 1 else (ot[0] if ot else None),
            "LG": og[0] if og else None, "RG": og[1] if len(og) > 1 else (og[0] if og else None),
            "C": self._nth("C", 0, out),
        }

    def _build_defense(self, nickel: bool = False, out: set | None = None) -> dict[str, dict]:
        d = {
            "EDGE1": self._nth("EDGE", 0, out), "EDGE2": self._nth("EDGE", 1, out),
            "DT1": self._nth("DT", 0, out), "DT2": self._nth("DT", 1, out),
            "ILB1": self._nth("ILB", 0, out), "ILB2": self._nth("ILB", 1, out),
            "CB1": self._nth("CB", 0, out), "CB2": self._nth("CB", 1, out),
            "S1": self._nth("S", 0, out), "S2": self._nth("S", 1, out),
        }
        if nickel:
            d["CB3"] = self._nth("CB", 2, out)
        return d

    def kicker(self) -> dict | None:
        return self._nth("K", 0)


@functools.lru_cache(maxsize=64)
def roster(team: str) -> Roster:
    return Roster(team, load_pool()[team])


def team_list() -> list[str]:
    return sorted(load_pool())

File: src/engine/roster.ts
/**
 * Team rosters, depth charts, per-play lineups (spec §14, §15).
 *
 * TS port of `analysis/engine/roster.py`. Depth-chart *order* uses `overall`
 * (allowed — it only ranks players into roles); play *outcomes* use the
 * individual skill attributes via the rating layer (`ratings.ts`).
 */

import type { Player } from "../schema/player.js";
import { loadPlayerPool } from "../data/players.js";

export const OFF_SLOTS = [
  "QB1", "RB1", "WR1", "WR2", "WR3", "TE1", "LT", "LG", "C", "RG", "RT",
] as const;
export const DEF_SLOTS_BASE = [
  "EDGE1", "EDGE2", "DT1", "DT2", "ILB1", "ILB2", "CB1", "CB2", "S1", "S2",
] as const;
export const DEF_SLOTS_NICKEL = [
  "EDGE1", "EDGE2", "DT1", "DT2", "ILB1", "CB1", "CB2", "CB3", "S1", "S2",
] as const;

export type Lineup = Record<string, Player | null>;

function pushInto<K, V>(m: Map<K, V[]>, k: K, v: V): void {
  const list = m.get(k);
  if (list) list.push(v);
  else m.set(k, [v]);
}

let _pool: Map<string, Player[]> | null = null;

/** Pool grouped by `nfl_team`; free agents ("FA" / blank) dropped. Memoised. */
export function loadPool(): Map<string, Player[]> {
  if (_pool) return _pool;
  const byTeam = new Map<string, Player[]>();
  for (const p of loadPlayerPool()) {
    const t = p.nfl_team;
    if (!t || t === "FA") continue;
    pushInto(byTeam, t, p);
  }
  _pool = byTeam;
  return byTeam;
}

export class Roster {
  readonly team: string;
  /** position -> players, sorted by `overall` descending */
  readonly depth: Map<string, Player[]>;
  private _off: Lineup | null = null;
  private readonly _def = new Map<boolean, Lineup>();

  constructor(team: string, players: Player[]) {
    this.team = team;
    this.depth = new Map();
    for (const p of players) pushInto(this.depth, p.position, p);
    for (const list of this.depth.values()) {
      list.sort((a, b) => (b.overall ?? 0) - (a.overall ?? 0));
    }
  }

  /**
   * nth-*available* player at `pos`, clamped to the last, or null if none.
   * `out` (ids injured/unavailable) is skipped before counting to `i`.
   */
  private nth(pos: string, i: number, out?: ReadonlySet<string>): Player | null {
    let d = this.depth.get(pos) ?? [];
    if (out && out.size) {
      const avail = d.filter((p) => !out.has(p.id));
      if (avail.length) d = avail;
    }
    return d[i] ?? d[d.length - 1] ?? null;
  }

  /** `out` = ids to treat as unavailable (in-game injuries). Uncached when non-empty. */
  offense(out?: ReadonlySet<string>): Lineup {
    if (out && out.size) return this.buildOffense(out);
    if (!this._off) this._off = this.buildOffense();
    return this._off;
  }

  defense(nickel = false, out?: ReadonlySet<string>): Lineup {
    if (out && out.size) return this.buildDefense(nickel, out);
    let d = this._def.get(nickel);
    if (!d) {
      d = this.buildDefense(nickel);
      this._def.set(nickel, d);
    }
    return d;
  }

  kicker(): Player | null {
    return this.nth("K", 0);
  }

  punter(): Player | null {
    return this.nth("P", 0);
  }

  private buildOffense(out?: ReadonlySet<string>): Lineup {
    let ot = this.depth.get("OT") ?? [];
    let og = this.depth.get("OG") ?? [];
    if (out && out.size) {
      ot = ot.filter((p) => !out.has(p.id));
      og = og.filter((p) => !out.has(p.id));
    }
    return {
      QB1: this.nth("QB", 0, out),
      RB1: this.nth("RB", 0, out),
      WR1: this.nth("WR", 0, out),
      WR2: this.nth("WR", 1, out),
      WR3: this.nth("WR", 2, out),
      TE1: this.nth("TE", 0, out),
      LT: ot[0] ?? null,
      RT: ot[1] ?? ot[0] ?? null,
      LG: og[0] ?? null,
      RG: og[1] ?? og[0] ?? null,
      C: this.nth("C", 0, out),
    };
  }

  private buildDefense(nickel: boolean, out?: ReadonlySet<string>): Lineup {
    const d: Lineup = {
      EDGE1: this.nth("EDGE", 0, out),
      EDGE2: this.nth("EDGE", 1, out),
      DT1: this.nth("DT", 0, out),
      DT2: this.nth("DT", 1, out),
      ILB1: this.nth("ILB", 0, out),
      ILB2: this.nth("ILB", 1, out),
      CB1: this.nth("CB", 0, out),
      CB2: this.nth("CB", 1, out),
      S1: this.nth("S", 0, out),
      S2: this.nth("S", 1, out),
    };
    if (nickel) d.CB3 = this.nth("CB", 2, out);
    return d;
  }
}

const _rosters = new Map<string, Roster>();

export function roster(team: string): Roster {
  let r = _rosters.get(team);
  if (!r) {
    const players = loadPool().get(team);
    if (!players) throw new Error(`roster: no players for team ${team}`);
    r = new Roster(team, players);
    _rosters.set(team, r);
  }
  return r;
}

export function teamList(): string[] {
  return [...loadPool().keys()].sort();
}
